kv/kvnemesis: TestKVNemesisMultiNode_BufferedWritesNoLockDurabilityUpgrades failed [txnrecovery during refresh retry]

kv/kvnemesis.TestKVNemesisMultiNode_BufferedWritesNoLockDurabilityUpgrades [failed](https://mesolite.cluster.engflow.com/invocations/default/dc9f0886-3136-4738-b80f-4e860a8313bf?testReportRun=3&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZuZW1lc2lzOmt2bmVtZXNpc190ZXN0) on master @ [8387e9ccb0960b5c207df74f8a8abcc979715eeb](https://github.com/cockroachdb/cockroach/commits/8387e9ccb0960b5c207df74f8a8abcc979715eeb):


```
    kvnemesis_test.go:468: seed: 3420669331701251985
    test_server_shim.go:594: DRPC is enabled (override by TestingGlobalDRPCOption)
    kvnemesis_test.go:259: kvnemesis logging to /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_3_of_3/test.outputs/kvnemesis3258629657
    kvnemesis.go:182: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.Snapshot)
          txn.SetUserPriority(roachpb.UserPriority(0.001000))
          txn.SetBufferedWritesEnabled(true)
          txn.GetForUpdateSkipLockedGuaranteedDurability(ctx, tk(5658748251140747936)) // @1759305822.257779417,0 (<nil>, <nil>)
          txn.Get(ctx, tk(7187098675379421789)) // @1759305822.257779417,0 (<nil>, <nil>)
          txn.ScanForShareSkipLocked(ctx, tk(2137898750724371111), tk(13975849236034048140), 0) // @1759305822.257779417,0 <nil>
          txn.ScanForUpdateSkipLocked(ctx, tk(5392635058644518597), tk(16254768641004895565), 0) // @1759305822.257779417,0 <nil>
          txn.ReverseScanForUpdateSkipLocked(ctx, tk(7809147440933845881), tk(17928732165494968149), 0) // @1759305822.257779417,0 <nil>
          txn.ReverseScanForShare(ctx, tk(13129801675006618678), tk(14381122015921834248), 0) // @1759305822.257779417,0 <nil>
          txn.ReverseScanForShareGuaranteedDurability(ctx, tk(8680778924895486362), tk(12180145692592774201), 0) // @1759305822.257779417,0 <nil>
          txn.DelMustAcquireExclusiveLock(ctx, tk(10753434680438621290) /* @s1 */) // @1759305822.257779417,0 <nil>
          txn.ScanForUpdate(ctx, tk(6451204819582027971), tk(11359773872436326519), 0) // @1759305822.257779417,0 <nil>
          txn.Put(ctx, tk(3581013804327173398), sv(2)) // <nil>
          txn.ScanForShareSkipLocked(ctx, tk(3908697965680049602), tk(17334379257185338308), 0) // @1759305822.257779417,0 (/Table/100/"4c8bffa2eb8ffaf2":v41, /Table/100/"de1f6ad20de01779":v5, <nil>)
          b := &kv.Batch{}
          b.DelMustAcquireExclusiveLock(tk(3581013804327173398) /* @s3 */) // transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED)
          b.CPutAllowIfDoesNotExist(tk(13498465455948326784), sv(4), exp(non-existent value)) // transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED)
          txn.CommitInBatch(ctx, b) // transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED)
          return nil
        }) // transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED): transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED)
    kvnemesis.go:182: error applying x.DelMustAcquireExclusiveLock(ctx, tk(3581013804327173398) /* @s3 */) // transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED): transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED)
    kvnemesis.go:182: error applying x.CPutAllowIfDoesNotExist(ctx, tk(13498465455948326784), sv(4), exp(non-existent value)) // transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED): transaction unexpectedly committed, ba: Delete [/Table/100/"31b25114feb0fd16"], ConditionalPut [/Table/100/"bb542ed072ae6780"], EndTxn(parallel commit) [/Table/100/"4e87e98d969b72a0"], [txn: bd494ed0]. txn: "unnamed" meta={id=bd494ed0 key=/Table/100/"4e87e98d969b72a0" iso=Snapshot pri=0.00000000 epo=0 ts=1759305822.882076064,1 min=1759305822.257779417,0 seq=5} lock=true stat=COMMITTED rts=1759305822.882076064,1 gul=1759305822.267779417,0 obs={n1@1759305822.401912606,0} int=5 ifw=2: TransactionStatusError: already committed (REASON_TXN_COMMITTED)
    kvnemesis.go:182: uncommitted snapshot txn had writes: [d]/Table/100/"953bdfde22e9046a":1759305822.882076064,1-><nil>@s1 [w]/Table/100/"31b25114feb0fd16":missing->v2@s2 [d]/Table/100/"31b25114feb0fd16":1759305822.882076064,1-><nil>@s3 [w]/Table/100/"bb542ed072ae6780":1759305822.882076064,1->v4@s4
    kvnemesis.go:223: failures(verbose): /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_3_of_3/test.outputs/kvnemesis3258629657/failures
        repro steps: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_3_of_3/test.outputs/kvnemesis3258629657/repro.go
        rangefeed KVs: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_3_of_3/test.outputs/kvnemesis3258629657/kvs-rangefeed.txt
        scan KVs: /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_3_of_3/test.outputs/kvnemesis3258629657/kvs-scan.txt
    kvnemesis_test.go:507: [0] proposed: 2650
    kvnemesis_test.go:509: [0] reproposed unchanged: 78
    kvnemesis_test.go:511: [0] reproposed with new LAI: 9
    kvnemesis_test.go:507: [1] proposed: 0
    kvnemesis_test.go:509: [1] reproposed unchanged: 0
    kvnemesis_test.go:511: [1] reproposed with new LAI: 0
    kvnemesis_test.go:507: [2] proposed: 0
    kvnemesis_test.go:509: [2] reproposed unchanged: 0
    kvnemesis_test.go:511: [2] reproposed with new LAI: 0
    kvnemesis_test.go:516: 
        	Error Trace:	pkg/kv/kvnemesis/kvnemesis_test.go:516
        	            				pkg/kv/kvnemesis/kvnemesis_test.go:365
        	Error:      	Should be zero, but was 4
        	Test:       	TestKVNemesisMultiNode_BufferedWritesNoLockDurabilityUpgrades
        	Messages:   	kvnemesis detected failures
    dump_raft_logs.go:32: dumping raft logs to /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_3_of_3/test.outputs/kvnemesis3258629657/raftlogs
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestKVNemesisMultiNode_BufferedWritesNoLockDurabilityUpgrades1043231307
--- FAIL: TestKVNemesisMultiNode_BufferedWritesNoLockDurabilityUpgrades (32.98s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode_BufferedWritesNoLockDurabilityUpgrades.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-54971

--- a/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_server_test.go
@@ -4410,6 +4410,7 @@ func TestUnexpectedCommitOnTxnRecovery(t *testing.T) {
 	targetTxnIDString.Store("")
 	ctx := context.Background()
 	st := cluster.MakeTestingClusterSettings()
+
 	// This test relies on unreplicated locks to be replicated on lease transfers.
 	concurrency.UnreplicatedLockReliabilityLeaseTransfer.Override(ctx, &st.SV, true)
 	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{
@@ -4625,7 +4626,7 @@ func TestMaxSpanRequestKeysWithMixedReadWriteBatches(t *testing.T) {
 //
 // Setup:
 //
-//	Create a split at key=8 to force batch splitting.
+//	Create a split at key=6 to force batch splitting.
 //	Write to key=8 to be able to observe if the subsequent delete was effective.
 //
 // Txn 1:
@@ -4756,4 +4757,171 @@ func TestRollbackAfterRefreshAndFailedCommit(t *testing.T) {
 	// programming error. So, we check the trace for it.
 	recording := collectAndFinish()
 	require.NotContains(t, recording.String(), "failed indeterminate commit recovery: programming error")
+}
+
+// TestUnexpectedCommitOnTxnAbortAfterRefresh is a regression test for #151864.
+// It is similar to TestRollbackAfterRefreshAndFailedCommit but tests a much
+// worse outcome that can be experienced by weak-isolation transactions.
+//
+// This test issues requests across two transactions. In the presence of the
+// bug, Txn 1 ends being committed even though an error is returned to the
+// user.
+//
+// Setup:
+//
+//	Create a split at key=6 to force batch splitting.
+//
+// Txn 1:
+//
+//	Read key=1 to set timestamp and prevent server side refresh
+//
+// Txn 2:
+//
+//	Read key=1 to bump timestamp cache on key 1
+//	Write key=8
+//
+// Txn1:
+//
+//	Batch{
+//	  Del key=1
+//	  Del key=2
+//	  Del key=8
+//	  Del key=9
+//	  EndTxn
+//	}
+//
+// This batch should encounter a write too old error on key=8.
+//
+// After a successful refresh, our testing filters have arranged for the second
+// EndTxn to fail. In the absence of the bug, we expect for this EndTxn failure
+// to result in the transaction never committing and an error being returned to
+// the client. When the bug existed, transaction recovery (either initiated by
+// another transaction or initiated by Txn 1 itself during a rollback issued
+// after the injected error) could result in the transaction being erroneously
+// committed despite the injected error being returned to the client.
+func TestUnexpectedCommitOnTxnAbortAfterRefresh(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	var (
+		targetTxnIDString atomic.Value
+		firstEndTxnSeen   atomic.Bool
+		cmdID             atomic.Value
+	)
+	cmdID.Store(kvserverbase.CmdIDKey(""))
+	targetTxnIDString.Store("")
+	ctx := context.Background()
+	st := cluster.MakeClusterSettings()
+	kvcoord.RandomizedTxnAnchorKeyEnabled.Override(ctx, &st.SV, false)
+
+	s, _, db := serverutils.StartServer(t, base.TestServerArgs{
+		Settings: st,
+		Knobs: base.TestingKnobs{
+			Store: &kvserver.StoreTestingKnobs{
+				TestingProposalFilter: func(fArgs kvserverbase.ProposalFilterArgs) *kvpb.Error {
+					if fArgs.Req.Header.Txn == nil ||
+						fArgs.Req.Header.Txn.ID.String() != targetTxnIDString.Load().(string) {
+						return nil // not our txn
+					}
+
+					t.Logf("proposal from our txn: %s", fArgs.Req)
+					endTxnReq := fArgs.Req.Requests[len(fArgs.Req.Requests)-1].GetEndTxn()
+					if endTxnReq != nil {
+						if !firstEndTxnSeen.Load() {
+							firstEndTxnSeen.Store(true)
+						} else {
+							epoch := fArgs.Req.Header.Txn.Epoch
+							t.Logf("will fail application for txn %s@epoch=%d; req: %+v; raft cmdID: %s",
+								fArgs.Req.Header.Txn.ID.String(), epoch, endTxnReq, fArgs.CmdID)
+							cmdID.Store(fArgs.CmdID)
+						}
+					}
+					return nil
+				},
+				TestingApplyCalledTwiceFilter: func(fArgs kvserverbase.ApplyFilterArgs) (int, *kvpb.Error) {
+					if fArgs.CmdID == cmdID.Load().(kvserverbase.CmdIDKey) {
+						t.Logf("failing application for raft cmdID: %s", fArgs.CmdID)
+						return 0, kvpb.NewErrorf("test injected error")
+					}
+					return 0, nil
+				},
+			},
+		},
+	})
+	defer s.Stopper().Stop(ctx)
+
+	scratchStart, err := s.ScratchRange()
+	require.NoError(t, err)
+
+	scratchKey := func(idx int) roachpb.Key {
+		key := scratchStart.Clone()
+		key = append(key, []byte(fmt.Sprintf("key-%03d", idx))...)
+		return key
+	}
+
+	_, _, err = s.SplitRange(scratchKey(6))
+	require.NoError(t, err)
+
+	tracer := s.TracerI().(*tracing.Tracer)
+	txn1Ctx, collectAndFinish := tracing.ContextWithRecordingSpan(context.Background(), tracer, "txn1")
+
+	txn1Err := db.Txn(txn1Ctx, func(txn1Ctx context.Context, txn1 *kv.Txn) error {
+		txn1.SetDebugName("txn1")
+		targetTxnIDString.Store(txn1.ID().String())
+		// Txn1 must be at either SNAPSHOT or READ COMMITTED with Stepping enabled
+		// for this bug because it both needs to be in an isolation level that
+		// allows committing when wrt != rts and the read needs to produce a read
+		// span that requires a refresh such that we can't do a server-side refresh.
+		_ = txn1.ConfigureStepping(txn1Ctx, kv.SteppingEnabled)
+		if err := txn1.SetIsoLevel(isolation.ReadCommitted); err != nil {
+			return err
+		}
+		if _, err = txn1.Get(txn1Ctx, scratchKey(1)); err != nil {
+			return err
+		}
+
+		// Txn2 now executes, arranging for the WriteTooOld error and the timestamp
+		// cache bump on the txn's anchor key.
+		if txn1.Epoch() == 0 {
+			txn2Ctx := context.Background()
+			txn2 := db.NewTxn(txn2Ctx, "txn2")
+			require.NoError(t, err)
+			_, err = txn2.Get(txn2Ctx, scratchKey(1))
+			require.NoError(t, err)
+
+			err = txn2.Put(txn2Ctx, scratchKey(8), "hello")
+			require.NoError(t, err)
+			err = txn2.Commit(txn2Ctx)
+			require.NoError(t, err)
+		}
+
+		b := txn1.NewBatch()
+		b.Del(scratchKey(1))
+		b.Del(scratchKey(2))
+		b.Del(scratchKey(8))
+		b.Del(scratchKey(9))
+		return txn1.CommitInBatch(txn1Ctx, b)
+	})
+	val8, err := db.Get(ctx, scratchKey(8))
+	require.NoError(t, err)
+
+	defer func() {
+		recording := collectAndFinish()
+		if t.Failed() {
+			t.Logf("TXN 1 TRACE: %s", recording)
+		}
+	}()
+
+	if val8.Exists() {
+		// If val8 _exists_ then it means our transaction did not commit. So really
+		// any error is correct.
+		require.Error(t, txn1Err)
+	} else {
+		// If val8 doesn't exist, then txn1 was committed so we shouldn't have
+		// gotten an error or se should have gotten an ambiguous result error.
+		if txn1Err != nil {
+			ambigErr := &kvpb.AmbiguousResultError{}
+			require.ErrorAs(t, txn1Err, &ambigErr, "transaction committed but non-ambiguous error returned")
+		}
+	}
 }

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go
@@ -260,6 +260,21 @@ func (sr *txnSpanRefresher) maybeRefreshAndRetrySend(
 	if !ok {
 		return nil, pErr
 	}
+
+	// If we are in a weak isolation transaction and we had an EndTxn in our
+	// batch, then this retry faces another hazard: The STAGING record written in
+	// the first attempt may possibly be satisfied by our future writes.
+	//
+	// TODO(ssd): If we were guaranteed to have a BatchResponse if a request was
+	// evaluated, then we could narrow this further and only bump the refresh
+	// timestamp if the StagingTimestamp == RefreshTS.
+	endTxnArg, hasET := ba.GetArg(kvpb.EndTxn)
+	bumpedRefreshTimestampRequired := hasET && txn.Status == roachpb.STAGING && txn.IsoLevel.ToleratesWriteSkew()
+	if bumpedRefreshTimestampRequired {
+		refreshTS = refreshTS.Next()
+		log.Eventf(ctx, "bumping refresh timestamp to avoid unexpected parallel commit: %s", refreshTS)
+	}
+
 	refreshFrom := txn.ReadTimestamp
 	refreshToTxn := txn.Clone()
 	refreshToTxn.BumpReadTimestamp(refreshTS)
@@ -267,9 +282,9 @@ func (sr *txnSpanRefresher) maybeRefreshAndRetrySend(
 	case roachpb.PENDING:
 	case roachpb.STAGING:
 		// If the batch resulted in an error but the EndTxn request succeeded,
-		// staging the transaction record in the process, downgrade the status
-		// back to PENDING. Even though the transaction record may have a status
-		// of STAGING, we know that the transaction failed to implicitly commit.
+		// staging the transaction record in the process, downgrade the status back
+		// to PENDING. Even though the transaction record may have a status of
+		// STAGING, we know that the transaction failed to implicitly commit.
 		refreshToTxn.Status = roachpb.PENDING
 	default:
 		return nil, kvpb.NewError(errors.AssertionFailedf(
@@ -294,8 +309,7 @@ func (sr *txnSpanRefresher) maybeRefreshAndRetrySend(
 
 	// To prevent starvation of batches and to ensure the correctness of parallel
 	// commits, split off the EndTxn request into its own batch on auto-retries.
-	args, hasET := ba.GetArg(kvpb.EndTxn)
-	if len(ba.Requests) > 1 && hasET && !args.(*kvpb.EndTxnRequest).Require1PC {
+	if len(ba.Requests) > 1 && hasET && !endTxnArg.(*kvpb.EndTxnRequest).Require1PC {
 		log.Eventf(ctx, "sending EndTxn separately from rest of batch on retry")
 		return sr.splitEndTxnAndRetrySend(ctx, ba)
 	}

